pallets/elections/src/impls.rs

```diff
@@ -293,6 +299,45 @@ where
 
         T::ValidatorRewardsHandler::add_rewards(rewards);
     }
+
+    fn calculate_underperforming_validators() {
+        let thresholds = CurrentEraCommitteeKickOutThresholds::<T>::get();
+        let current_committee = T::SessionInfoProvider::current_committee();
+        for validator in current_committee {
+            let underperformance = match SessionValidatorBlockCount::<T>::try_get(&validator) {
+                Ok(block_count) => block_count <= thresholds.block_count_threshold,
+                Err(_) => true,
+            };
+            if underperformance {
+                Self::mark_validator_underperformance(&thresholds, &validator)
+            }
+        }
+    }
+
+    fn mark_validator_underperformance(
+        thresholds: &CommitteeKickOutThresholds,
+        validator: &T::AccountId,
+    ) {
+        UnderperformedValidatorSessionCount::<T>::mutate(&validator, |count| {
+            *count += 1;
+        });
+        let counter = UnderperformedValidatorSessionCount::<T>::get(&validator);
+        if counter >= thresholds.underperformed_session_count_threshold {
+            // ideally we'd include here information about how many sessions a validator missed,
+            // yet in no_std world String does not exist. Instead BoundedVec, we should use some
+            // encoded error struct
+            let reason: Vec<u8> = "Insufficient uptime".into();
```

Comment:
This is possible, but we would need one variant to be `Custom(BoundedVec<...>)` anyway, because we sometimes want custom kicks by sudo. From the perspective of frontend it's better to just have strings here, to avoid complication...

Reply:
Actually I merged both remarks - now `KickOutReason` is an `enum`

Reply:
cool, leaving this conversation for @DamianStraszak to resolve :)